pkg/mic/mic.go

```diff
@@ -259,26 +273,39 @@ func (c *Client) Sync(exit <-chan struct{}) {
 					continue
 				}
 
-				glog.V(5).Infof("Initiating assigned id creation for pod - %s, binding - %s", createID.Spec.Pod, binding.Name)
-
-				err = c.createAssignedIdentityDeps(&createID, id, node)
-				if err != nil {
-					// Since k8s event has only Info and Warning, using Warning.
-					c.EventRecorder.Event(binding, corev1.EventTypeWarning, "binding apply error",
-						fmt.Sprintf("Applying binding %s node %s for pod %s resulted in error %v", binding.Name, createID.Spec.NodeName, createID.Name, err))
-					glog.Error(err)
-					continue
+				if isUserAssignedMSI {
+					c.appendToAddListForNode(nodeMap, id.Spec.ResourceID, node)
 				}
+
 				appliedBinding := createID.Spec.AzureBindingRef
-				eventRecordStart := time.Now()
 				glog.V(5).Infof("Binding applied: %+v", appliedBinding)
-				c.EventRecorder.Event(appliedBinding, corev1.EventTypeNormal, "binding applied",
-					fmt.Sprintf("Binding %s applied on node %s for pod %s", appliedBinding.Name, createID.Spec.NodeName, createID.Name))
-				stats.Update(stats.EventRecord, time.Since(eventRecordStart))
 			}
-			stats.Put(stats.TotalIDAdd, time.Since(beginAdding))
 		}
 
+		// one final createorupdate to each node in the map
+		// TODO: parallelize this process
+		var nodeWithErrors []string
+		for _, n := range nodeMap {
+			beginAdding := time.Now()
+
+			// generate unique list so we don't make multiple calls to assign/remove same id
+			n.addUserAssignedMSIIDs = c.getUniqueIDs(n.addUserAssignedMSIIDs)
+			n.removeUserAssignedMSIIDs = c.getUniqueIDs(n.removeUserAssignedMSIIDs)
+
+			err = c.CloudClient.AppendOrRemoveUserMSI(n.addUserAssignedMSIIDs, n.removeUserAssignedMSIIDs, n.node)
+			if err != nil {
+				// check which all identity assignment failed
+				// remove those assigned identities
+				// TODO check with identity team if CreateOrUpdate results in error, is it all or some failed
```

Comment:
Can we do this before the PR gets merged.

Reply:
Sent out an email.